ivre/db/__init__.py

```diff
@@ -2467,16 +2465,135 @@ def _get_timeslots(cls, start_time, end_time):
         Returns an array of timeslots included between start_time and end_time
         """
         times = []
-        time = cls.date_round(start_time)
-        end_timeslot = cls.date_round(end_time)
-        while time <= end_timeslot:
+        first_timeslot = cls._get_timeslot(
+            start_time, config.FLOW_TIME_PRECISION, config.FLOW_TIME_BASE
+        )
+        time = first_timeslot['start']
+        last_timeslot = cls._get_timeslot(
+            end_time, config.FLOW_TIME_PRECISION, config.FLOW_TIME_BASE
+        )
+        end_time = last_timeslot['start']
+        while time <= end_time:
             d = OrderedDict()
             d['start'] = time
             d['duration'] = config.FLOW_TIME_PRECISION
             times.append(d)
             time += timedelta(seconds=config.FLOW_TIME_PRECISION)
         return times
 
+    @staticmethod
+    def _get_timeslot(time, precision, base):
+        ts = utils.datetime2timestamp(time)
+        ts += utils.tz_offset(ts)
+        new_ts = ts - (((ts % precision) - base) % precision)
+        new_ts -= utils.tz_offset(new_ts)
+        d = OrderedDict()
+        d["start"] = datetime.fromtimestamp(new_ts)
+        d["duration"] = precision
+        return d
+
+    def reduce_precision(self, new_duration, flt=None,
+                         base=None, before=None, after=None, precision=None):
+        """
+        Changes precision of timeslots to <new_duration> precision of flows
+        honoring:
+            - the given filter <flt> if specified
+            - that have been seen before <before> if specified
+            - that have been seen after <after> if specified
+            - timeslots changed must currently have <precision> precision if
+                specified
+        <base> represents the timestamp of the base point.
+        If <precision> is specified:
+            <new_duration> must be a multiple of <precision>
+            <new_duration> must be greater than <precision>
+            <base> must be a multiple of current precision
```

Comment:
Actually `base` should not have been kept here because it has been removed from the MongoDB implementation. Instead, I get the base from the config (`FLOW_TIME_BASE`). Otherwise, I would have had to support multiple bases, leading to store in each timeslot its base. It seems like an unnecessary overhead.